pkg/plugins/runtime/k8s/controllers/cni_taint_controller.go

```diff
@@ -0,0 +1,188 @@
+	log := r.Log.WithValues("node", req.NamespacedName)
+
+	// Fetch the Node instance
+	node := &kube_core.Node{}
+	if err := r.Get(ctx, req.NamespacedName, node); err != nil {
+		if kube_apierrs.IsNotFound(err) {
+			log.Error(err, "node not found")
+			return kube_ctrl.Result{}, nil
+		}
+		log.Error(err, "unable to fetch node")
+		return kube_ctrl.Result{}, err
+	}
+	log.Info("node successfully fetched")
+
+	kubeSystemPods := &kube_core.PodList{}
+	namespaceOption := kube_client.InNamespace(cniPodNamespace)
+	matchingFields := kube_client.MatchingFields{nodeIndexField: node.Name}
+	matchingLabels := kube_client.MatchingLabels{cniAppLabel: r.CniApp}
+	if err := r.Client.List(ctx, kubeSystemPods, namespaceOption, matchingFields, matchingLabels); err != nil {
+		return kube_ctrl.Result{}, err
+	}
+
+	err := r.updateTaints(ctx, log, node, kubeSystemPods.Items)
+	if err != nil {
+		return kube_ctrl.Result{}, errors.Wrap(err, "unable to update node taints")
+	}
+
+	return kube_ctrl.Result{}, err
+}
+
+func (r *CniNodeTaintReconciler) updateTaints(ctx context.Context, log logr.Logger, node *kube_core.Node, pods []kube_core.Pod) error {
+	taintIndex := slices.IndexFunc(node.Spec.Taints, func(taint kube_core.Taint) bool {
+		return taint.Key == nodeReadinessTaintKey && taint.Effect == kube_core.TaintEffectNoSchedule
+	})
+
+	if taintIndex >= 0 {
+		if r.hasCniPodRunning(log, pods) {
+			log.Info("has cni pod running and taint")
+			return r.untaintNode(ctx, log, node, taintIndex)
+		} else {
+			log.Info("has no cni pod running and taint")
+			return nil
+		}
+	} else {
+		if r.hasCniPodRunning(log, pods) {
+			log.V(1).Info("has cni pod running and no taint")
+			return nil
+		} else {
+			log.Info("has no cni pod running and no taint")
+			return r.taintNode(ctx, log, node)
+		}
+	}
+}
+
+func (r *CniNodeTaintReconciler) untaintNode(ctx context.Context, log logr.Logger, node *kube_core.Node, taintIndex int) error {
+	node.Spec.Taints = slices.Delete(node.Spec.Taints, taintIndex, taintIndex+1)
+
+	err := r.Client.Update(ctx, node)
+	if err == nil {
+		log.Info("removed the taint from node")
+	}
+	return err
+}
+
+func (r *CniNodeTaintReconciler) taintNode(ctx context.Context, log logr.Logger, node *kube_core.Node) error {
+	node.Spec.Taints = append(node.Spec.Taints, kube_core.Taint{
+		Key:    nodeReadinessTaintKey,
+		Effect: kube_core.TaintEffectNoSchedule,
+	})
+
+	err := r.Client.Update(ctx, node)
+	if err == nil {
+		log.Info("added taint to node")
+	}
+	return err
+}
+
+func (r *CniNodeTaintReconciler) hasCniPodRunning(log logr.Logger, pods []kube_core.Pod) bool {
+	for _, pod := range pods {
+		isCniPod := pod.Labels[cniAppLabel] == r.CniApp
+		if isCniPod && pod.Status.Phase == kube_core.PodRunning {
+			for _, condition := range pod.Status.Conditions {
+				if condition.Type == kube_core.PodReady && condition.Status == kube_core.ConditionTrue {
```

Comment:
I think this code can be improved with some of the [`apimachinery` util functions](https://pkg.go.dev/k8s.io/apimachinery/pkg/api/meta#IsStatusConditionTrue)

Reply:
https://pkg.go.dev/k8s.io/apimachinery/pkg/api/meta#IsStatusConditionTrue is not built for this type, it uses [Condition](https://pkg.go.dev/k8s.io/apimachinery@v0.24.3/pkg/apis/meta/v1#Condition) and not [PodCondition](https://pkg.go.dev/k8s.io/api@v0.24.3/core/v1#PodCondition)